sympy/matrices/__init__.py

```diff
@@ -25,11 +26,28 @@
     MatrixSlice, BlockDiagMatrix, BlockMatrix, FunctionMatrix, Identity,
     Inverse, MatAdd, MatMul, MatPow, MatrixExpr, MatrixSymbol, Trace,
     Transpose, ZeroMatrix, OneMatrix, blockcut, block_collapse, matrix_symbols, Adjoint,
-    hadamard_product, HadamardProduct, HadamardPower, Determinant, det,
+    hadamard_product, HadamardProduct, HadamardPower, Determinant,
     diagonalize_vector, DiagMatrix, DiagonalMatrix, DiagonalOf, trace,
     DotProduct, kronecker_product, KroneckerProduct,
     PermutationMatrix, MatrixPermute)
 
+from .determinant import (
+    adjugate, charpoly, cofactor, cofactor_matrix,
+    det, det_bareiss, det_berkowitz, det_LU, minor, minor_submatrix)
+
+from .reductions import (
+    is_echelon, echelon_form, rank, rref)
```

Comment:
Of these I would only export `rank` I think and that can also make sense for `MatrixSymbol`.

Reply:
So you want me to remove the top level `is_echelon`, `echelon_form` and `rref` completely or only available through `from sympy.matrices import rref`?

Have a look at how I did `matrices.expressions.determinant.det()` and let me know if you want to do it like this. This would also work for `adjugate`, `charpoly` and `rank` when they get implemented, until then should I export those from `sympy.matrices`?

So when you figure out which ones you want to make it to top level sympy import give me the list and I add it.

Reply:
What function is used to calculate the determinant here:
```julia
In [7]: A = MatrixSymbol('A', 2, 2)                                                                                                            

In [8]: B = Matrix([[1, 2], [3, 4]])                                                                                                           

In [9]: det(A)                                                                                                                                 
Out[9]: │A│

In [10]: det(A).subs(A, B)                                                                                                                     
Out[10]: 
│⎡1  2⎤│
│⎢    ⎥│
│⎣3  4⎦│

In [11]: det(A).subs(A, B).doit()                                                                                                              
Out[11]: -2
```
It should be the same as when calling `det(B)`.

Reply:
In each case a `Determinant` is expression is created but the call to `det_matrix` only happens on the `doit`. Are you saying it should happen on the `subs`?

Reply:
If you mean which `det` function gets control it is always `matrices.expressions.determinant.det()` first.

Reply:
I think that one case calls `M._eval_determinant` and the other calls `det_matrix(M)`.

Reply:
Same underlying function, `._eval_determinant` is `matrices.determinants._det` which is wrapped by `matrices.determinants.det_matrix`.

Reply:
Actually I think `._eval_determinant` can be removed and just use `.det`, it only seems to be called from `Determinant.doit` anyway.